kasa/discover.py

```diff
@@ -315,5 +312,64 @@ def _get_device_class(info: dict) -> Type[SmartDevice]:
                 return SmartLightStrip
 
             return SmartBulb
+        raise UnsupportedDeviceException("Unknown device type: %s" % type_)
+
+    @staticmethod
+    def _get_device_instance_legacy(data: bytes, ip: str, port: int) -> SmartDevice:
+        """Get SmartDevice for device described by passed data from legacy 9999 response."""
+        try:
+            info = json_loads(TPLinkSmartHomeProtocol.decrypt(data))
+        except Exception as ex:
+            raise SmartDeviceException(
+                f"Unable to read response from device: {ip}: {ex}"
+            )
+
+        _LOGGER.debug("[DISCOVERY] %s << %s", ip, info)
+
+        device_class = Discover._get_device_class(info)
+        device = device_class(ip, port=port)
+        device.update_from_discover_info(info)
+        return device
 
-        raise SmartDeviceException("Unknown device type: %s" % type_)
+    @staticmethod
+    def _get_device_instance(
+        data: bytes, ip: str, port: int, credentials: Credentials
+    ) -> SmartDevice:
+        """Get SmartDevice for device described by passed data from new 20002 response."""
+        try:
+            info = json_loads(data[16:])
+        except Exception as ex:
+            raise SmartDeviceException(
+                f"Unable to read response from device: {ip}: {ex}"
+            )
+
+        if (
+            "result" in info
+            and "mgt_encrypt_schm" in info["result"]
+            and "encrypt_type" in info["result"]["mgt_encrypt_schm"]
+            and info["result"]["mgt_encrypt_schm"]["encrypt_type"] == "KLAP"
+            and "lv" not in info["result"]["mgt_encrypt_schm"]
+            and "device_type" in info["result"]
+        ):
```

Comment:
As we depend already on pydantic, it's probably best to create a `BaseModel` for the discovery info. This also avoids the need to have nested dict accesses for these tests here.